### PR TITLE
Font-family in WASM

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): #915 

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Custom fonts don't work on WASM, as code attempts to set the CSS `font-family` value to the exact XAML value (e.g. `ms-appx:///Assets/Fonts/MyFont.ttf#My Font`).

## What is the new behavior?

- Font name is now parsed from the XAML FontFamily value to work well with WASM.
- Docs are updated to explain how the value is parsed and how it matches CSS font-faces
- `FontFamily.Source` still returns the original value to stay in line with UWP, new internal `ParsedSource` property on WASM contains the parsed version
- Added unit tests for font source parsing

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

### CODE REVIEW

File: doc/articles/using-uno-ui.md [hunk 1]
```diff
@@ -387,14 +387,46 @@ Adding a custom font is done through the use of WebFonts, using a data-URI:
...
+Will match the following `@font-face`:
+
+```css
+@font-face {
+  font-family: "Roboto";
+  ...
+}
+```
+
+In case your `FontFamily` value does not contain `#`, we fall back to the font file name. Hence for:
```
Comment (jeromelaban) [inline doc/articles/using-uno-ui.md]:
```suggestion
In case your `FontFamily` value does not contain `#`, Uno fall back to the font file name. Hence for:
```

File: doc/articles/using-uno-ui.md [hunk 2]
```diff
@@ -387,14 +387,46 @@ Adding a custom font is done through the use of WebFonts, using a data-URI:
 
 ```css
 @font-face {
-  font-family: "Symbols";
+  font-family: "Roboto";
```
Comment (jeromelaban) [inline doc/articles/using-uno-ui.md]: Why did you change the font family ?

File: src/Uno.UI/UI/Xaml/FontFamily.cs [hunk 3]
```diff
@@ -10,6 +11,10 @@ public FontFamily(string familyName)
 		{
 			Source = familyName;
 
+#if __WASM__
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/FontFamily.cs]: Should be moved to a `FontFamily.wasm.cs` file and initialized through a partial method.

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/FontFamily.cs]: Done!

File: doc/articles/using-uno-ui.md (hunk 2 above)
Comment (MartinZikmund) [reply doc/articles/using-uno-ui.md]: Ah! Originally with good intentions - I wanted to show the custom font matching there directly, so I wanted to use well-known font name. But then I split it to make it clearer... And didn't roll this one back 😀

File: doc/articles/using-uno-ui.md (hunk 1 above)
Comment (MartinZikmund) [reply doc/articles/using-uno-ui.md]: Fixd'

**Comment by MartinZikmund** [comment]
@jeromelaban Would it be okay if I move the Custom fonts section fron `using-uno-ui.md` to a separate file and put it among the **Features** section of docs? Because right now it is a bit buried and not as discoverable (I used repo search to find it in the first place).

**Comment by davidjohnoliver** [comment]
> @jeromelaban Would it be okay if I move the **Custom fonts** section fron `using-uno-ui.md` to a separate file and put it among the **Features** group of docs? Because right now it is a bit buried and not as discoverable (I used repo search to find it in the first place).

Please do! Was just thinking about that today. Just make sure to leave the existing section header in `using-uno-ui.md` with a link under it to the new file (to not break peoples' bookmarks).

**Comment by MartinZikmund** [comment]
> > @jeromelaban Would it be okay if I move the **Custom fonts** section fron `using-uno-ui.md` to a separate file and put it among the **Features** group of docs? Because right now it is a bit buried and not as discoverable (I used repo search to find it in the first place).
> 
> Please do! Was just thinking about that today. Just make sure to leave the existing section header in `using-uno-ui.md` with a link under it to the new file (to not break peoples' bookmarks).

Done, good thinking with the link 👌

**Comment by carldebilly** [comment]
Philosophical question: should we add something about the potential legal implications of embedding a foreign font into an app?

**Comment by MartinZikmund** [comment]
> Philosophical question: should we add something about the potential legal implications of embedding a foreign font into an app?

Not sure if that would stop people from doing that anyway though,.. We could suggest some sources of nice free fonts - Google Fonts for example, plus add some quick guide on how to create a custom icon font using Icons8/NoonProject + icomoon or similar tools.

File: src/Uno.UI/UI/Xaml/FontFamily,wasm.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+﻿#if __WASM__
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/FontFamily,wasm.cs]: You don't need this `#if` when the file is using the `.wasm.cs` prefix...

Ooohhh...  your file is using a comma instead of a dot!!

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/FontFamily,wasm.cs]: 🤣🤣🤣🤣

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/FontFamily,wasm.cs]: We in Czech Republic like our commas 😂

**Comment by MartinZikmund** [comment]
@dr1rrb if you are on-line, can you help me get this merged? Planning a blog post re custom fonts in Uno for Thursday, so it would be great to have WASM ready too 😊

File: doc/articles/features/custom-fonts.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
+#### Custom Fonts on Android
```
Comment (dr1rrb) [inline doc/articles/features/custom-fonts.md]: You should add a title to this file, and remove few "#" for this sub-title :)

Comment (MartinZikmund) [reply doc/articles/features/custom-fonts.md]: Fix'd :-)

Comment (MartinZikmund) [reply doc/articles/features/custom-fonts.md]: Thanks a lot for the quick approve!

Comment (dr1rrb) [reply doc/articles/features/custom-fonts.md]: You're welcome :) I've put the ready-to-merge flag, but I don't know if it works with external contributor, otherwise I'll try to merge it later tonight.

**Comment by MartinZikmund** [comment]
@dr1rrb oh, that build failed randomly, can you please reset it?

**Comment by dr1rrb** [comment]
https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=11540&view=results hopefully it will pass ... we have a lot of issue recently with the CI :/

**Comment by MartinZikmund** [comment]
@dr1rrb awesome, it worked! Thanks!
